### PR TITLE
Basics of view profunctor

### PR DESCRIPTION
**Pull request by mbrandonw**
Just wanna get this in place so that we can start messing with it.

There's a new type `View<P, N>`, which is just a wrapper for functions `P -> [N]` (think props to array of nodes). This is of course a monoid and profunctor, so I've implemented all of the functions and operators.

there are also two helper functions. one called `fold` which allows us to apply the view to props, but then flatten to a single node (instead of an array) by wrapping in a parent `div`. there's also a help for lifting any node into a view by making it a `Unit -> [Node]` constant function.

### CODE REVIEW

File: Tests/HTMLTests/ViewTests.swift [hunk 1]
```diff
@@ -0,0 +1,103 @@
+import XCTest
+import Prelude
+import HTML
+import CSS
+
+class ViewTests: XCTestCase {
+  func testSemigroupAssociativity() {
```
Comment (mbrandonw) [inline Tests/HTMLTests/ViewTests.swift]: this is technically "semigroup associativity with respect to `fold`".

we already know `View` is a semigroup by the fact that it is the set of functions into a semigroup, but we wanna make sure that our "`fold`" is invariant under the semigroup operator.

File: Sources/HTML/Reader.swift [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+  public func contramap<S>(_ f: @escaping (S) -> R) -> Reader<S, A> {
+    return .init(f >>> self.runReader)
+  }
+
+  public func flatMap<B>(_ f: @escaping (A) -> Reader<R, B>) -> Reader<R, B> {
+    return .init { r in f(self.runReader(r)).runReader(r) }
+  }
+
+  public static func pure(_ a: A) -> Reader<R, A> {
```
Comment (stephencelis) [inline Sources/HTML/Reader.swift]: We should just overload all the things! Interested in seeing what some of this looks like with operators.

File: Sources/HTML/View.swift [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+public func >¢< <A, B, N>(f: @escaping (A) -> B, view: View<B, N>) -> View<A, N> {
+  return view.contramap(f)
+}
+
+// MARK: Helpers
+
+extension View where N == Node {
+  // Views into HTML nodes carry an additional structure that allows us to
+  // fold props into a single node by wrapping them in a parent div.
+  public func fold(_ props: P) -> N {
```
Comment (stephencelis) [inline Sources/HTML/View.swift]: Interested to see if we need this. Seems like rendering directly or folding to our array of HTML elements may be good enough.

**Comment by mbrandonw** [comment]
ok I just pushed a change that got rid of `fold` for generating a node, but it does add a `.rendered(props, config)` method so that we don't have to do `render(view.call)`. maybe we'll remove that someday too, but just from the perspective of tests it seemed nicer. hopefully real world experience will show how this API should evolve.

I also made `render([nodes])` which makes it a monoid morphism

**Pull request merged**